### PR TITLE
[SPARK-34444][SQL] Pushdown scalar-subquery filter to FileSourceScan

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr pushdown scalar-subquery filter to FileSourceScan. For example:
```scala
sql("CREATE TABLE t1 using parquet AS SELECT id AS a, id AS b FROM range(500000000L)")
sql("CREATE TABLE t2 using parquet AS SELECT id AS d FROM range(20)")
sql("SELECT * FROM t1 WHERE b = (select max(d) from t2)").show
```
Before this pr | After this pr
-- | --
image | image

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by peter-toth** [comment]
@wangyum, how does this PR relates to https://github.com/apache/spark/pull/23802?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala [hunk 1]
```diff
@@ -85,6 +85,26 @@ object SubqueryExpression {
...
+      case _ => false
+    }.isDefined
+  }
+
+  /**
+   * Returns true when an expression contains a non-scalar subquery
+   */
+  def hasNonScalarSubquery(e: Expression): Boolean = {
+    e.find {
+      case s: SubqueryExpression if !hasScalarSubquery(s) => true
```
Comment (peter-toth) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:103]: how about `case s: SubqueryExpression => !s.isInstanceOf[ScalarSubquery]` and we probably don't need `def hasScalarSubquery()` at all?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 2]
```diff
@@ -335,6 +336,19 @@ case class FileSourceScanExec(
...
+  private lazy val runtimePushedDownFilters = {
+    dataFilters.flatMap {
+      case e: Expression if ExecSubqueryExpression.hasScalarSubquery(e) =>
+        val updatedValue = e.transform {
+          case s: ScalarSubquery => s.value
+        }
+        Some(updatedValue)
+      case _ =>
+        Nil
+    }.flatMap(translateFilter(_, DataSourceUtils.supportNestedPredicatePushdown(relation)))
```
Comment (peter-toth) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:349]: nit: IMO we can combine the 2 `flatMap`s and the code is still easy to read

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 3]
```diff
@@ -166,14 +167,11 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
...
-      // subquery expressions are filtered out because they can't be used to prune buckets or pushed
-      // down as data filters, yet they would be executed
-      val normalizedFiltersWithoutSubqueries =
-        normalizedFilters.filterNot(SubqueryExpression.hasSubquery)
-
       val bucketSpec: Option[BucketSpec] = fsRelation.bucketSpec
       val bucketSet = if (shouldPruneBuckets(bucketSpec)) {
-        genBucketSet(normalizedFiltersWithoutSubqueries, bucketSpec.get)
+        // subquery expressions are filtered out because they can't be used to prune buckets
+        // or pushed down as data filters, yet they would be executed
```
Comment (peter-toth) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:173]: nit: `or pushed down as data filters` is not valid here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 4]
```diff
@@ -184,7 +182,9 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
       // Partition keys are not available in the statistics of the files.
       // `dataColumns` might have partition columns, we need to filter them out.
       val dataColumnsWithoutPartitionCols = dataColumns.filterNot(partitionColumns.contains)
-      val dataFilters = normalizedFiltersWithoutSubqueries.flatMap { f =>
+      // Non-scalar subquery expressions are filtered out because they can't be used to prune
+      // buckets or pushed down as data filters, yet they would be executed
```
Comment (peter-toth) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:186]: bi: `to prune buckets` is not valid here

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala [hunk 5]
```diff
@@ -85,6 +85,26 @@ object SubqueryExpression {
...
+    e.find {
+      case _: ScalarSubquery => true
+      case _ => false
+    }.isDefined
+  }
+
+  /**
+   * Returns true when an expression contains a non-scalar subquery
+   */
+  def hasNonScalarSubquery(e: Expression): Boolean = {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:101]: Isn't it equal to `hasSubquery(e) && !hasScalarSubquery(e)`?

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by wangyum** [comment]
Thank you all. I need to fix the failed test.

**Comment by SaurabhChawla100** [comment]
@wangyum - Just curious to clear my understanding about this change.
I am able to see the scalar sub-query in the example without joins.
`select max(d) from t2)`

Are we pushing down the scalar sub-query which is having  single or multiple joins and than returning the single row 
For eg 
`SELECT * FROM t1 WHERE b = (select t2.id from t2 , t3 , t4  where t2.id = t3.id1 and t2.date = t4.date1)`

So now if the `select t2.id from t2 , t3 , t4  where t2.id = t3.id1 and t2.date = t4.date1` this takes time to return the result 

if we pushing the complex query , then we are holding the scan of the table t1 until the subquery gets completed and pushdown is done on t1 and after that it starts processing.

Will it not impact the performance compare to what is the existing functionality where both table scans starts at same time.

Second point what if the table t1 is having less number of rows do we still want to push down the scalar subquery which is the complex query itself.

**Comment by wangyum** [comment]
Thank you @SaurabhChawla100.

**Comment by kmozaid** [comment]
@SaurabhChawla100 Hi Saurabh, I didn't get your statement `If Scalar Subquery completes first, than only scan of t1 starts before this change and after this PR also , Than push down of scalar subquery will always be faster`

The query `SELECT * FROM t1 WHERE b = (select max(d) from t2);` given in this example completes very fast and still it is not being pushed down. 

Suggestion - Can we provide hint to planner to execute subquery first and push down subquery filter?

cc @wangyum Could you please reopen this PR?

**Pull request closed without merging**